src/Registration/cReg/NiftiImageData.cpp

```diff
@@ -64,27 +64,64 @@ static void check_folder_exists_if_not_create(const std::string &path)
 template<class dataType>
 NiftiImageData<dataType>::NiftiImageData(const NiftiImageData<dataType>& to_copy)
 {
-    *this = dynamic_cast<const ImageData&>(to_copy);
+    *this = dynamic_cast<const ImageData<dataType>&>(to_copy);
 }
 
 template<class dataType>
 NiftiImageData<dataType>& NiftiImageData<dataType>::operator=(const NiftiImageData<dataType>& to_copy)
 {
-    *this = dynamic_cast<const ImageData&>(to_copy);
+    *this = dynamic_cast<const ImageData<dataType>&>(to_copy);
     return *this;
 }
 
 template<class dataType>
-NiftiImageData<dataType>::NiftiImageData(const ImageData& to_copy)
+NiftiImageData<dataType>::NiftiImageData(const ImageData<float>& to_copy)
 {
     *this = to_copy;
 }
 
 template<class dataType>
-NiftiImageData<dataType>& NiftiImageData<dataType>::operator=(const ImageData& to_copy)
+NiftiImageData<dataType>::NiftiImageData(const ImageData<complex_float_t>& to_copy)
+{
+    *this = to_copy;
+}
+
+template<class dataType>
+NiftiImageData<dataType>& NiftiImageData<dataType>::operator=(const ImageData<float>& to_copy)
 {
     // Check for self-assignment
-    if (this != &to_copy) {
+    if (static_cast<void*>(this) != static_cast<const void*>(&to_copy)) {
+        // Try to cast to NiftiImageData.
+        const NiftiImageData<dataType> * const nii_ptr = dynamic_cast<const NiftiImageData<dataType> * const >(&to_copy);
```

Comment:
no idea, this is Richard's stuff, and it looks like he wanted to cover for lots of data types - from `bool` to `long double` (see `NiftiImageData.h`)

at any rate, I doubt we would ever need any `dataType` other than `float`

to remind you: `dataType` cannot be complex - complex images are treated as pairs of real ones

Reply:
Checking `copy_nifti_image` it contains https://github.com/SyneRBI/SIRF/blob/60bda8a52414a77172825594ddc78c484321122e/src/Registration/cReg/NiftiImageData.cpp#L683
therefore, it needs to have the same data-types, see these lines are correct.

Note that it can be simplified to
```
auto nii_ptr = dynamic_cast<const NiftiImageData<dataType> * const >(&to_copy)
```

Reply:
> therefore, it needs to have the same data-types

so far as I can see, this is to be taken care of by `nifti_copy_nim_info`

> Note that it can be simplified

thanks, adopted